Add public blog (#12)
aicoe-aiops/categorical-encoding: docs/blog/blog.md

#### Impact

Red Hat collects systems data from RHEL and Openshift subscribers. A lot of these data points are nominal categorical such as the CPU model name: _Intel Xeon CPU E5-2643_, _Intel Xeon Platinum 8175M_, _AMD Opteron Processor 6176 SE_. Other examples include categorical variables in Linux configuration files, product names, alert names, and various log files keys. The artifacts derived from these experiments can directly benefit unsupervised problems with these datasets.

## Encoding

In this section, we describe various types of encoders that we experimented with. For each encoder, we briefly describe the process of encoding and also their pros and cons in terms of performance and explainability. For a better explanation, we are using an example categorical variable employee title that could have values such as software developer, civil engineer, etc.

### Base N Encoders

Base N Encoders directly convert the categories into numbers. We show three different methods:

##### Label / Integer Encoding

- Process
  - Base = N
  - Convert categories to integers

>   - For example, all Civil Engineer I will be represented by integer k

Reviewer: ```suggestion
  - For example, all _Civil Engineer I_ will be represented by integer _k_
```

- Pros
  - Fast and saves memory as the output vector is in a smaller dimension
  - Can be used for multiple categorical variables at the same time
- Cons
  - Collisions happen if the output vector size is small
  - Inverse mapping is not available so no interpretation

### Word Based Encoders

#### Language Model: Fasttext Encoder [[4](https://arxiv.org/abs/1607.04606)]

- Process
  - Downloads a pre-trained language model on English words
  - For each category, performs a forward run of the pre-trained model
  - Outputs the vectors from the last layer of the model
- Pros
  - Dynamic length of the output vector
  - Captures meaning in the words or categories
- Cons
  - Heavy file size of the pre-trained model
  - May not work well with words that don't appear a lot in natural English
  - Output vector dimensions have no explainability

##### Similarity Encoder [[5](https://dirty-cat.github.io/stable/generated/dirty_cat.SimilarityEncoder.html)]

- Process
  - Create one column per category
  - Define n-gram similarity
  - For each category, compare similarity with other categories
  - Instead of 0 and 1 in one-hot encoding use similarity value

>   - E.g. Categories: Associate Software Developer, Software Developer, Senior Software Developer

Reviewer: ```suggestion
  - E.g. Categories: _Associate Software Developer_, _Software Developer_, _Senior Software Developer_
```

- Pros
  - In high cardinality, categorical variables different entries are often variations on the same entities
  - Using simple one-hot encoding will create orthogonal features, whereas it is clear that those 3 terms have a lot in common.

>   - If we wanted to use word embedding methods such as word2vec, we would have to go through a cleaning phase: those algorithms are not trained to work on data such as ‘Accountant/Auditor I’. However, this can be error-prone and time-consuming.

Reviewer: ```suggestion
  - If we wanted to use word embedding methods such as word2vec, we would have to go through a cleaning phase: those algorithms are not trained to work on data such as _Accountant/Auditor I_. However, this can be error-prone and time-consuming.
```

>   - The high dimensions can be meaningfully reduced to lower dimensions: All to Software developers, similar to clustering.

Reviewer: ```suggestion
  - The high dimensions can be meaningfully reduced to lower dimensions: _All_ to _Software developer_s, similar to clustering.
```

- Cons
  - High dimensional output vector
  - Works only with variables that have meaningful substring overlaps

##### Online Gamma Poisson Factorization [[6](https://arxiv.org/pdf/1907.01860.pdf)]

- Process
  - Estimates a decomposition of the string entries in terms of a linear combination of latent categories.

>   - Similar to LDA for finding topics in documents but since the string entries here are much shorter than text documents and can contain typos, it uses ngrams level representation.

Reviewer: ```suggestion
  - Similar to [LDA](https://en.wikipedia.org/wiki/Latent_Dirichlet_allocation) for finding topics in documents but since the string entries here are much shorter than text documents and can contain typos, it uses ngrams level representation.
```

![Results](./results.png "Results")

Figure 3: Weighted average f1 scores of encoding schemes

## Application

Next, we are going to see how we can apply gamma poisson factorization in an unsupervised setting.

![Dataset](./dataset.png "Data set")

Figure 4: Data set of system hardware configurations

- **Data:** We use a data set of hardware configuration of user systems. Figure 4 shows the system configurations.

- **Problem:** Our aim is to find configurations that are outliers and rules that determine that these configurations are outliers.

- **Encoding:** We use gamma poisson factorization for the variable _modelname_ since it has a high cardinality. For the rest, we use one-hot encoding since they have a low cardinality.

> - **Outlier detection:** To find the outliers, we use the isolation forest algorithm that finds points that are easiest to isolate.

Reviewer: ```suggestion
- **Outlier detection:** To find the outliers, we use the [isolation forest](https://en.wikipedia.org/wiki/Isolation_forest) algorithm that finds points that are easiest to isolate.
```

or 
```suggestion
- **Outlier detection:** To find the outliers, we use the [isolation forest](https://cs.nju.edu.cn/zhouzh/zhouzh.files/publication/tkdd11.pdf) algorithm that finds points that are easiest to isolate.
```

- **Interpretation:** Finally, we train a surrogate decision tree based on the labels given by the isolation forest to find rule paths that lead to the outlier classification.

#### **Interpretation**

![Rulepath](./rulepath.png "Rule path")

Figure 5: Example rule path for detecting a system as an anomaly

##### The tree can be interpreted in the following way:

- The nodes represent feature splitting points.
- The first line represents a condition used to split that node.
- The second line specifies the percentage of samples that fall in that node.
- The third line specifies that, for those samples that fall in that node, what percent belong to class Anomaly (orange color) and what % belongs to class Not Anomaly (Blue color)

##### Rule paths

> - One-hot encoded variable: if a node says that vendor_seabios &lt;= 0.5, that means that for all the systems where the vendor is not seabios. Since one-hot encoding means 0 or 1, &lt; 0.5 means 0.

Reviewer: ```suggestion
- One-hot encoded variable: if a node says that `vendor_seabios <= 0.5`, that means that for all the systems where the vendor is not _seabios_. Since one-hot encoding means 0 or 1, &lt; 0.5 means 0.
```

> - Gammas Poisson factorized variables: if a node says modelname_broadwell,skylake,haswell &lt;= R (real number), then that means that systems that have strings broadwell, skylake or haswell in their modelname are important factors for splitting the tree further. In order to decide whether their presence or absence causes these anomalies, further validation has to be done.

Reviewer: ```suggestion
- Gammas Poisson factorized variables: if a node says `modelname_broadwell,skylake,haswell <= R` (real number), then that means that systems that have strings _broadwell_, _skylake_ or _haswell_ in their _modelname_ are important factors for splitting the tree further. In order to decide whether their presence or absence causes these anomalies, further validation has to be done.
```

> - Example rule path found in Figure 4: if (modelname has platinum, pentium, 8175m) and (vendor is seabios) then it could be an anomaly.

Reviewer: ```suggestion
- Example rule path found in Figure 4: if (modelname has _platinum_, _pentium_, _8175m_) and (vendor is _seabios_) then it could be an anomaly.
```
